### PR TITLE
Fix/ping

### PR DESCRIPTION
**Pull request by jannyHou**
connect to strongloop/loopback-connector-cloudant#175
Test case modified in PR https://github.com/strongloop/loopback-connector-couchdb2/pull/33

### CODE REVIEW

File: lib/cloudant.js [hunk 1]
```diff
@@ -114,5 +114,28 @@ Cloudant.prototype.getModelObjectSettings = function(mo) {
...
+      if (err) cb('ping failed');
+      self.getDriverInst().db.list(function(err, result) {
+        returnCB(err, result);
+      });
+    });
+  }
+
+  function returnCB(err, result) {
+    debug('Cloudant.prototype.ping results %j %j', err, result);
+    if (err) cb('ping failed');
```
Comment (b-admike) [inline lib/cloudant.js]: would `if (err) cb(new Error('ping failed'))` work better here?

**Comment by jannyHou** [comment]
@kjdelisle we already have `ping` succeeds, for `ping` fails, other than an invalid `url`, I am not sure what I can add, any idea?

File: lib/cloudant.js (hunk 1 above)
Comment (jannyHou) [reply lib/cloudant.js]: ah true, sorry I copied the old code

**Comment by jannyHou** [comment]
@kjdelisle The test case is added in PR strongloop/loopback-connector-couchdb2#33, hope it passes CI and it should be checked in before this Cloudant PR, let's put the comment in that PR first :-p

File: lib/cloudant.js [hunk 2]
```diff
@@ -114,5 +114,28 @@ Cloudant.prototype.getModelObjectSettings = function(mo) {
...
+  debug('Cloudant.prototype.ping');
+  var self = this;
+  var driverInst = self.getDriverInst();
+  if (driverInst) {
+    driverInst.db.list(function(err, result) {
+      returnCB(err, result);
+    });
+  } else {
+    self.dataSource.once('connected', function(err) {
+      if (err) cb('ping failed');
```
Comment (kjdelisle) [inline lib/cloudant.js]: `cb(new Error('ping failed'))`

Comment (kjdelisle) [reply lib/cloudant.js]: To match the same pattern below.

Comment (jannyHou) [reply lib/cloudant.js]: @kjdelisle Fixed :)

**Comment by jannyHou** [comment]
@slnode test please

**Comment by jannyHou** [comment]
@slnode test please

**Comment by jannyHou** [comment]
@kjdelisle The ["fail" test case(checking `ping failed`)](https://github.com/strongloop/loopback-connector-couchdb2/blob/master/test/connection.test.js#L30-L31) won't work for cloudant, since event `connected` won't happen: https://github.com/strongloop/loopback-connector-cloudant/pull/177/files#diff-fa06ecf3249c63aabcbf7ff1ae8ec186R126

The error is 
> Error: connect ECONNREFUSED 127.0.0.1:4

instead of Error object `ping failed`

Are you ok if I disable that test case for cloudant? I don't think you can test it.

**Comment by kjdelisle** [comment]
@jannyHou Are you saying it's a huge amount of effort to _override_ the test? If the expected response is not identical to the CouchDB response, it makes sense to reimplement the test to be Cloudant-specific, but I don't agree that we should just disable all forms of test coverage for the ping function.

**Comment by jannyHou** [comment]
@kjdelisle I can write a test for it, but it's an uncaught error, not from cb

**Comment by kjdelisle** [comment]
@jannyHou Is this a side-effect of the cloudant driver not correctly providing the error via callback/promise?

**Comment by jannyHou** [comment]
@kjdelisle The uncaught is not even returned by `ping`, it gets generated when you do a `var db = getDataSource(fakeConfig)`, I don't think it's a driver's problem, the juggler is written in that way to return error, see test case in mongodb: https://github.com/strongloop/loopback-connector-mongodb/blob/master/test/mongodb.test.js#L35

The interesting part is I am not sure why the couchdb doesn't fail when calling `getDataSource` with a fake config.

**Comment by kjdelisle** [comment]
@jannyHou Then that's the logic you should imitate for the cloudant version of the test.

**Comment by kjdelisle** [comment]
@jannyHou As long as the tests pass (and are there!) I'm happy.

File: test/cloudant.connection.test.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+// Copyright IBM Corp. 2017. All Rights Reserved.
+// Node module: loopback-connector-couchdb2
```
Comment (Amir-61) [inline test/cloudant.connection.test.js]: `loopback-connector-couchdb2` -> `loopback-connector-cloudant`

File: lib/cloudant.js [hunk 4]
```diff
@@ -114,5 +114,28 @@ Cloudant.prototype.getModelObjectSettings = function(mo) {
...
+  if (driverInst) {
+    driverInst.db.list(function(err, result) {
+      returnCB(err, result);
+    });
+  } else {
+    self.dataSource.once('connected', function(err) {
+      if (err) cb(new Error('ping failed'));
+      self.getDriverInst().db.list(function(err, result) {
+        returnCB(err, result);
+      });
```
Comment (Amir-61) [inline lib/cloudant.js]:
```
      self.getDriverInst().db.list(function(err, result) {
         returnCB(err, result);
       });
```
can be changed to:
```
     self.getDriverInst().db.list(returnCB(err, result));
```
to me It is much cleaner and neater...

File: lib/cloudant.js [hunk 5]
```diff
@@ -114,5 +114,28 @@ Cloudant.prototype.getModelObjectSettings = function(mo) {
...
 };
 
+Cloudant.prototype.ping = function(cb) {
+  debug('Cloudant.prototype.ping');
+  var self = this;
+  var driverInst = self.getDriverInst();
+  if (driverInst) {
+    driverInst.db.list(function(err, result) {
+      returnCB(err, result);
+    });
```
Comment (Amir-61) [inline lib/cloudant.js]:
```
    driverInst.db.list(function(err, result) {
       returnCB(err, result);
     });
```
can be changed to:
```
     driverInst.db.list(returnCB(err, result));
```
@kjdelisle @jannyHou to me It is much cleaner and neater...

File: test/cloudant.connection.test.js (hunk 3 above)
Comment (kjdelisle) [reply test/cloudant.connection.test.js]: Missed that one, good catch

**Pull request merged**